File: merkle-trie/src/main/java/net/consensys/cava/trie/Node.java
package net.consensys.cava.trie;

import net.consensys.cava.bytes.Bytes;
import net.consensys.cava.bytes.Bytes32;
import net.consensys.cava.concurrent.AsyncResult;

import java.util.Optional;

interface Node<V> {

  AsyncResult<Node<V>> accept(NodeVisitor<V> visitor, Bytes path);

  AsyncResult<Bytes> path();

  AsyncResult<Optional<V>> value();

  Bytes rlp();

  Bytes rlpRef();

  Bytes32 hash();

  AsyncResult<Node<V>> replacePath(Bytes path);
}

File: merkle-trie/src/main/kotlin/net/consensys/cava/trie/experimental/Node.kt
package net.consensys.cava.trie.experimental

import net.consensys.cava.bytes.Bytes
import net.consensys.cava.bytes.Bytes32

internal interface Node<V> {

  suspend fun accept(visitor: NodeVisitor<V>, path: Bytes): Node<V>

  suspend fun path(): Bytes

  suspend fun value(): V?

  fun rlp(): Bytes

  fun rlpRef(): Bytes

  fun hash(): Bytes32

  suspend fun replacePath(path: Bytes): Node<V>
}
